### ISSUE TITLE
Keys in the data dictionary passed to blaze server cannot contain '/' (forward slash).

### ISSUE THREAD

**Issue by cowlicks** [ContinuumIO/blaze#1110]
Other weird characters break  things too.
On the server

```python
from blaze import Server, resource
path = 'lastfm_subset/A/A/A/TRAAAAW128F429D538.json'
r = {path: resource(path)}
Server(r).run()
```
on the client
```python
In [2]: from blaze import Data
In [3]: d = Data('blaze://localhost:6363')


  File <nofile>, line 2
      lastfm_subset/A/A/A/TRAAAAW128F429D538.json: {
                   ^

DataShapeSyntaxError: Invalid DataShape token

---------------------------------------------------------------------------
DataShapeSyntaxError                      Traceback (most recent call last)
<ipython-input-3-71c21141fe52> in <module>()
----> 1 d = Data('blaze://localhost:6363')

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/blaze/interactive.py in Data(data, dshape, name, fields, columns, schema, **kwargs)
     66         dshape = datashape.dshape(dshape)
     67     if not dshape:
---> 68         dshape = discover(data)
     69         types = None
     70         if isinstance(dshape.measure, Tuple) and fields:

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/multipledispatch/dispatcher.py in __call__(self, *args, **kwargs)
    161             self._cache[types] = func
    162         try:
--> 163             return func(*args, **kwargs)
    164 
    165         except MDNotImplementedError:

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/blaze/server/client.py in discover(c)
     99 @dispatch(Client)
    100 def discover(c):
--> 101     return c.dshape
    102 
    103 

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/blaze/server/client.py in dshape(self)
     88             raise ValueError("Bad Response: %s" % reason(response))
     89 
---> 90         return dshape(content(response).decode('utf-8'))
     91 
     92 

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/util.py in dshape(o)
     47         return o
     48     if isinstance(o, py2help._strtypes):
---> 49         ds = parser.parse(o, type_symbol_table.sym)
     50     elif isinstance(o, (coretypes.CType, coretypes.String,
     51                         coretypes.Record, coretypes.JSON,

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse(ds_str, sym)
    570     """
    571     dsp = DataShapeParser(ds_str, sym)
--> 572     ds = dsp.parse_datashape()
    573     # If no datashape could be found
    574     if ds is None:

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_datashape(self)
    139                 return coretypes.DataShape(option(ds))
    140         else:
--> 141             return self.parse_datashape_nooption()
    142 
    143     def parse_datashape_nooption(self):

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_datashape_nooption(self)
    162                 return coretypes.DataShape(dim, *dshape.parameters)
    163         # Try dtype
--> 164         dtype = self.parse_dtype()
    165         if dtype:
    166             return coretypes.DataShape(dtype)

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_dtype(self)
    305                     return None
    306         elif tok.id == lexer.LBRACE:
--> 307             return self.parse_struct_type()
    308         elif tok.id == lexer.LPAREN:
    309             return self.parse_funcproto_or_tuple_type()

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_struct_type(self)
    477         fields = self.parse_homogeneous_list(self.parse_struct_field, lexer.COMMA,
    478                                              'Invalid field in struct',
--> 479                                              trailing_sep=True) or []
    480         if self.tok.id != lexer.RBRACE:
    481             self.raise_error('Invalid field in struct')

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_homogeneous_list(self, parse_item, sep_tok_id, errmsg, trailing_sep)
     71         while item is not None:
     72             # Parse the <item>
---> 73             item = parse_item()
     74             if item is not None:
     75                 items.append(item)

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_struct_field(self)
    504             return None
    505         name = self.tok.val
--> 506         self.advance_tok()
    507         if self.tok.id != lexer.COLON:
    508             self.raise_error('Expected a ":" separating the field ' +

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in advance_tok(self)
     38                 # we need to request a new token from the lexer
     39                 if self.pos >= len(self.tokens):
---> 40                     self.tokens.append(next(self.lex))
     41             except StopIteration:
     42                 # Create an EOF token, whose span starts at the

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/lexer.py in lex(ds_str)
     95             raise error.DataShapeSyntaxError(pos, '<nofile>',
     96                                              ds_str,
---> 97                                              'Invalid DataShape token')
     98         # Skip whitespace
     99         m = _whitespace_re.match(ds_str, pos)

DataShapeSyntaxError: 

  File <nofile>, line 2
      lastfm_subset/A/A/A/TRAAAAW128F429D538.json: {
                   ^

DataShapeSyntaxError: Invalid DataShape token


```

**Comment by cowlicks** [comment]
Note that this works when I replace the slashes with spaces via `path.replace('/', ' ')`.

**Comment by mrocklin** [comment]
If this is related to #1108 then I would recommend turning nested directories into nested dicts rather than having them all flat in the same dict.  Blaze knows how to traverse down dicts so we can probably keep the hierarchical structure.  

Outside of #1108 though this definitely seems like a valid error.  We probably need to escape/quote more heavily somewhere.

**Comment by cpcloud** [comment]
The error is that `str(dshape)` doesn't necessarily produce a datashape that can be fed back in to the `datashape.dshape` function. We need to quote non Python identifiers in the repr of `Record` `DataShape` instances

**Comment by cpcloud** [comment]
i believe this is fixed by https://github.com/ContinuumIO/datashape/pull/147

**Comment by cpcloud** [comment]
@cowlicks can you confirm that ContinuumIO/datashape#147 fixed this issue?

**Comment by cowlicks** [comment]
It  still fails for filenames with `.` them. See here:
```python
In [1]: from blaze import Data

In [2]: d = Data('blaze://localhost:6363')


  File <nofile>, line 4
          TRAAAAW128F429D538.json: {
                            ^

DataShapeSyntaxError: Invalid DataShape token

---------------------------------------------------------------------------
DataShapeSyntaxError                      Traceback (most recent call last)
<ipython-input-2-71c21141fe52> in <module>()
----> 1 d = Data('blaze://localhost:6363')

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/blaze/interactive.py in Data(data, dshape, name, fields, columns, schema, **kwargs)
     66         dshape = datashape.dshape(dshape)
     67     if not dshape:
---> 68         dshape = discover(data)
     69         types = None
     70         if isinstance(dshape.measure, Tuple) and fields:

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/multipledispatch/dispatcher.py in __call__(self, *args, **kwargs)
    161             self._cache[types] = func
    162         try:
--> 163             return func(*args, **kwargs)
    164 
    165         except MDNotImplementedError:

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/blaze/server/client.py in discover(c)
     99 @dispatch(Client)
    100 def discover(c):
--> 101     return c.dshape
    102 
    103 

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/blaze/server/client.py in dshape(self)
     88             raise ValueError("Bad Response: %s" % reason(response))
     89 
---> 90         return dshape(content(response).decode('utf-8'))
     91 
     92 

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/util.py in dshape(o)
     47         return o
     48     if isinstance(o, py2help._strtypes):
---> 49         ds = parser.parse(o, type_symbol_table.sym)
     50     elif isinstance(o, (coretypes.CType, coretypes.String,
     51                         coretypes.Record, coretypes.JSON,

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse(ds_str, sym)
    570     """
    571     dsp = DataShapeParser(ds_str, sym)
--> 572     ds = dsp.parse_datashape()
    573     # If no datashape could be found
    574     if ds is None:

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_datashape(self)
    139                 return coretypes.DataShape(option(ds))
    140         else:
--> 141             return self.parse_datashape_nooption()
    142 
    143     def parse_datashape_nooption(self):

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_datashape_nooption(self)
    162                 return coretypes.DataShape(dim, *dshape.parameters)
    163         # Try dtype
--> 164         dtype = self.parse_dtype()
    165         if dtype:
    166             return coretypes.DataShape(dtype)

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_dtype(self)
    305                     return None
    306         elif tok.id == lexer.LBRACE:
--> 307             return self.parse_struct_type()
    308         elif tok.id == lexer.LPAREN:
    309             return self.parse_funcproto_or_tuple_type()

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_struct_type(self)
    477         fields = self.parse_homogeneous_list(self.parse_struct_field, lexer.COMMA,
    478                                              'Invalid field in struct',
--> 479                                              trailing_sep=True) or []
    480         if self.tok.id != lexer.RBRACE:
    481             self.raise_error('Invalid field in struct')

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_homogeneous_list(self, parse_item, sep_tok_id, errmsg, trailing_sep)
     71         while item is not None:
     72             # Parse the <item>
---> 73             item = parse_item()
     74             if item is not None:
     75                 items.append(item)

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_struct_field(self)
    509                              'name from its datashape')
    510         self.advance_tok()
--> 511         ds = self.parse_datashape()
    512         if ds is None:
    513             self.raise_error('Expected the datashape of the field')

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_datashape(self)
    139                 return coretypes.DataShape(option(ds))
    140         else:
--> 141             return self.parse_datashape_nooption()
    142 
    143     def parse_datashape_nooption(self):

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_datashape_nooption(self)
    162                 return coretypes.DataShape(dim, *dshape.parameters)
    163         # Try dtype
--> 164         dtype = self.parse_dtype()
    165         if dtype:
    166             return coretypes.DataShape(dtype)

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_dtype(self)
    305                     return None
    306         elif tok.id == lexer.LBRACE:
--> 307             return self.parse_struct_type()
    308         elif tok.id == lexer.LPAREN:
    309             return self.parse_funcproto_or_tuple_type()

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_struct_type(self)
    477         fields = self.parse_homogeneous_list(self.parse_struct_field, lexer.COMMA,
    478                                              'Invalid field in struct',
--> 479                                              trailing_sep=True) or []
    480         if self.tok.id != lexer.RBRACE:
    481             self.raise_error('Invalid field in struct')

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_homogeneous_list(self, parse_item, sep_tok_id, errmsg, trailing_sep)
     71         while item is not None:
     72             # Parse the <item>
---> 73             item = parse_item()
     74             if item is not None:
     75                 items.append(item)

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_struct_field(self)
    509                              'name from its datashape')
    510         self.advance_tok()
--> 511         ds = self.parse_datashape()
    512         if ds is None:
    513             self.raise_error('Expected the datashape of the field')

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_datashape(self)
    139                 return coretypes.DataShape(option(ds))
    140         else:
--> 141             return self.parse_datashape_nooption()
    142 
    143     def parse_datashape_nooption(self):

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_datashape_nooption(self)
    162                 return coretypes.DataShape(dim, *dshape.parameters)
    163         # Try dtype
--> 164         dtype = self.parse_dtype()
    165         if dtype:
    166             return coretypes.DataShape(dtype)

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_dtype(self)
    305                     return None
    306         elif tok.id == lexer.LBRACE:
--> 307             return self.parse_struct_type()
    308         elif tok.id == lexer.LPAREN:
    309             return self.parse_funcproto_or_tuple_type()

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_struct_type(self)
    477         fields = self.parse_homogeneous_list(self.parse_struct_field, lexer.COMMA,
    478                                              'Invalid field in struct',
--> 479                                              trailing_sep=True) or []
    480         if self.tok.id != lexer.RBRACE:
    481             self.raise_error('Invalid field in struct')

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_homogeneous_list(self, parse_item, sep_tok_id, errmsg, trailing_sep)
     71         while item is not None:
     72             # Parse the <item>
---> 73             item = parse_item()
     74             if item is not None:
     75                 items.append(item)

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in parse_struct_field(self)
    504             return None
    505         name = self.tok.val
--> 506         self.advance_tok()
    507         if self.tok.id != lexer.COLON:
    508             self.raise_error('Expected a ":" separating the field ' +

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/parser.py in advance_tok(self)
     38                 # we need to request a new token from the lexer
     39                 if self.pos >= len(self.tokens):
---> 40                     self.tokens.append(next(self.lex))
     41             except StopIteration:
     42                 # Create an EOF token, whose span starts at the

/home/blake/.miniconda/envs/blaze3/lib/python3.4/site-packages/datashape/lexer.py in lex(ds_str)
     95             raise error.DataShapeSyntaxError(pos, '<nofile>',
     96                                              ds_str,
---> 97                                              'Invalid DataShape token')
     98         # Skip whitespace
     99         m = _whitespace_re.match(ds_str, pos)

DataShapeSyntaxError: 

  File <nofile>, line 4
          TRAAAAW128F429D538.json: {
                            ^

DataShapeSyntaxError: Invalid DataShape token

```

**Comment by cpcloud** [comment]
are you sure you're running datashape master?

**Comment by cpcloud** [comment]
```
In [1]: d = Data('blaze://localhost:6363')

In [2]: d
Out[2]:
Data:       <blaze.server.client.Client object at 0x114222bd0>
DataShape:  {
  '/Users/pcloud/data/trip/csv/all.csv': var * {
    medallion: ?string,
    hack_license: ?string,
    vendor_id: ?string,
    rate_code: int64,
    store_and_fwd_flag: ?string,
  ...

In [3]: d.dshape
Out[3]:
dshape("""{
  '/Users/pcloud/data/trip/csv/all.csv': var * {
    medallion: ?string,
    hack_license: ?string,
    vendor_id: ?string,
    rate_code: int64,
    store_and_fwd_flag: ?string,
    pickup_datetime: ?datetime,
    dropoff_datetime: ?datetime,
    passenger_count: int64,
    trip_time_in_secs: int64,
    trip_distance: ?float64,
    pickup_longitude: ?float64,
    pickup_latitude: ?float64,
    dropoff_longitude: ?float64,
    dropoff_latitude: ?float64,
    tolls_amount: ?float64,
    tip_amount: ?float64,
    total_amount: ?float64,
    mta_tax: ?float64,
    fare_amount: ?float64,
    payment_type: ?string,
    surcharge: ?float64
    }
  }""")
```

**Comment by cowlicks** [comment]
@cpcloud Oh! Sorry it does work. Thanks!

**Comment by cpcloud** [comment]
closed by ContinuumIO/datashape#147
